pkg/cloudclient/gcp/private.go

```diff
@@ -128,7 +130,7 @@ func (gc *Client) ensureDNSForService(kclient k8s.Client, svc *corev1.Service, d
 	}
 	rhapiLbIP := ingressList[0].IP
 	// ensure forwarding rule exists in GCP for service
-	err := gc.ensureGCPForwardingRuleForExtIP(rhapiLbIP)
+	err := gc.ensureGCPForwardingRuleForExtIP(rhapiLbIP, getForwardingRuleList)
```

Comment:
I'm not clear why we're passing a function as a parameter for this. Why can't `ensureGCPForwardingRuleForExtIp` call `getForwardingRuleList` directly?

Reply:
This follows higher order function mocking. It's one of the ways of mocking a function so that we cover precisely the function we're testing.  The higher order function is private and is kept extremely simple to isolate the external dependency. https://medium.com/@tahasevim231/mocking-with-high-order-functions-in-go-b76d5a3103d2

Reply:
Looking at this again, I'm not convinced getForwadingRuleList needs to be its own function at this point. Its only ever called once and its not a method of the gcp Client but instead takes it as an argument so I don't see much value in it being a function that's only available in this package as it current stands

Reply:
It's for mocking the unit test correctly. This is a common method of unit testing. It's increasing coverage, so it's good for the code overall.

Reply:
This changes the ergonomics of the code and makes it fairy complicated going forward but with no functional change, except to make mocking tests simpler. We have other functions with a similar pattern that would also all need to be refactored to this pattern if we want to test them as well.

IMO, we should instead abstract out the GCP API client from the gc Client to an interface with the list of required methods. That way we can pass in a mocked or "real" implementation to the gc Client struct.